ValveKeyValue/ValveKeyValue/Deserialization/KV1BinaryReader.cs

```diff
@@ -61,82 +61,81 @@ public void Dispose()
-                    value = new KVObjectValue<long>(reader.ReadInt64(), KVValueType.Int64);
-                    break;
+                var name = ReadNullTerminatedString();
+                if (type == KV1BinaryNodeType.ChildObject)
+                {
+                    listener.OnObjectStart(name);
+                    ReadObjectCore();
+                    listener.OnObjectEnd();
+                }
+                else
+                {
+                    KVValue value;
+
+                    switch (type)
+                    {
+                        case KV1BinaryNodeType.String:
+                            value = new KVObjectValue<string>(ReadNullTerminatedString(Encoding.UTF8), KVValueType.String);
+                            break;
+
+                        case KV1BinaryNodeType.WideString:
+                            throw new NotSupportedException("Wide String is not supported.");
+
+                        case KV1BinaryNodeType.Int32:
+                        case KV1BinaryNodeType.Color:
+                        case KV1BinaryNodeType.Pointer:
+                            value = new KVObjectValue<int>(reader.ReadInt32(), KVValueType.Int32);
+                            break;
+
+                        case KV1BinaryNodeType.UInt64:
+                            value = new KVObjectValue<ulong>(reader.ReadUInt64(), KVValueType.UInt64);
+                            break;
+
+                        case KV1BinaryNodeType.Float32:
+                            var floatValue = BitConverter.ToSingle(reader.ReadBytes(4), 0);
+                            value = new KVObjectValue<float>(floatValue, KVValueType.FloatingPoint);
+                            break;
+
+                        case KV1BinaryNodeType.Int64:
+                            value = new KVObjectValue<long>(reader.ReadInt64(), KVValueType.Int64);
+                            break;
+
+                        default:
+                            throw new ArgumentOutOfRangeException(nameof(type));
+                    }
 
-                default:
-                    throw new ArgumentOutOfRangeException(nameof(type));
+                    listener.OnKeyValuePair(name, value);
+                }
             }
-
-            listener.OnKeyValuePair(name, value);
         }
 
-        string ReadNullTerminatedString()
+        string ReadNullTerminatedString(Encoding encoding = null)
         {
-            var sb = new StringBuilder();
-            byte nextByte;
-            while ((nextByte = reader.ReadByte()) != 0)
-            {
-                sb.Append((char)nextByte);
-            }
+            if (encoding == null) encoding = Encoding.ASCII;
+            var b = reader.ReadByte();
 
-            return sb.ToString();
+            int len;
+            for (len = 0; b != 0; len++)
+                b = reader.ReadByte();
+
+            reader.BaseStream.Position -= len + 1;
+            var str = reader.ReadBytes(len);
```

Comment:
We don't need to read twice. Consider keeping a buffer, and resizing it if we overrun.

Reply:
Now using MemoryStream.